cpp/include/cugraph/prims/reduce_v.cuh

```diff
@@ -27,6 +27,33 @@
 namespace cugraph {
 namespace experimental {
 
+template <typename T>
+struct ValueAdd : public thrust::plus<T> {
+};
+
+template <typename... Args>
+struct ValueAdd<thrust::tuple<Args...>> : public thrust::binary_function<thrust::tuple<Args...>,
+                                                                         thrust::tuple<Args...>,
+                                                                         thrust::tuple<Args...>> {
+  using Type = thrust::tuple<Args...>;
+
+ private:
+  template <typename T, std::size_t... I>
+  __device__ constexpr auto sum_impl(T& t1, T& t2, std::index_sequence<I...>)
+  {
+    return thrust::make_tuple((thrust::get<I>(t1) + thrust::get<I>(t2))...);
+  }
+
+ public:
+  __device__ constexpr auto operator()(const Type& t1, const Type& t2)
+  {
+    return sum_impl(t1, t2, std::make_index_sequence<thrust::tuple_size<Type>::value>());
+  }
+};
+
+template <typename... Args>
+struct ValueAdd<std::tuple<Args...>> : public ValueAdd<thrust::tuple<Args...>> {
+};
```

Comment:
Why did you implement ValueAdd instead of reusing the existing one in https://github.com/rapidsai/cugraph/blob/branch-21.08/cpp/include/cugraph/prims/edge_op_utils.cuh?

Reply:
The existing functionality plus_edge_op_result is not a functor so it does not work directly with the thrust call here. I would have had to wrap it around another struct. The functor being used internally plus_thrust_tuple could be used though.

ValueAdd implementation was an attempt to modernize it a bit.

Reply:
Can't you just call ```plus_edge_op_result``` inside a functor? Why do we need ```sum_impl```?

And this functor should be used in transform_reduce_v as well, so we may place this functor elsewhere... (maybe edge_op_utils.cuh possibly renaming the file and functions to reflect that now these are used for both custom vertex & edge op outputs.

Reply:
And I may better not strongly recommend using a device lambda here (as device lambdas often fail to work properly for very esoteric reasons), but doing something like

```
[]__device__(auto lhs, auto rhs) {
  return plus_edge_op_result(lhs, rhs);
}
```

might be a simpler option (if this works... and I guess this will work here). If you want to avoid using device lambdas, then we may go for a simple wrapper functor than trying to reinvent the existing functionality.

Reply:
> Can't you just call plus_edge_op_result inside a functor? Why do we need sum_impl?

Yes, that's what I was referring to here : I would have had to wrap it around another struct.

My point was that I would just be increasing the number of abstraction levels. The new functor would wrap the function plus_edge_op_result, which wraps a functor plus_thrust_tuple. I was hoping to replace the existing calls to plus_edge_op_result (wherever applicable) to this since it is slightly more compact and modern since it does not use the recursive variadic implementation but index sequence instead.

And yes perhaps move it to some other generalized location would be a good idea.

Reply:
OK, then we should remove plus_edge_op_result rather than keeping both and make our codebase more redundant.